### PR TITLE
Improve field count typical case performance

### PR DESCRIPTION
**Pull request by runer112**
The tightest upper bound one can specify on the number of fields in a struct is `sizeof(type) * CHAR_BIT`. So this was previously used when performing a binary search for the field count. This upper bound is extremely loose when considering a typical large struct, which is more likely to contain a relatively small number of relatively large fields rather than the other way around. The binary search range being multiple orders of magnitude larger than necessary wouldn't have been a significant issue if each test was cheap, but they're not. Testing a field count of N costs O(N) memory and time. As a result, the initial few steps of the binary search may be prohibitively expensive.

The primary optimization introduced by these changes is to use unbounded binary search, a.k.a. exponential search, instead of the typically loosely bounded binary search. This produces a tight upper bound (within 2x) on the field count to then perform the binary search with.

As an upside of this change, the compiler-specific limit placed on the upper bound on the field count to stay within compiler limits could be removed.

### CODE REVIEW

**Comment by runer112** [comment]
This issue was originally noticed when some source files in a large project seemed to be consuming a suspiciously large amount of memory (and also time). After some digging, the culprit was eventually nailed down as field count detection.

Before these changes, compiling one such with clang 14 peaked at 1.5 GB of memory usage. After these changes, compiling the same file peaked at 617 MB. These numbers include the memory usage of compiling everything else as well, which is why the after-fix number is still relatively large. The memory usage attributable to field counting specifically probably went from something like 1 GB to something at least one or two orders of magnitude less.

**Comment by apolukhin** [comment]
The PR fails on msvc-14.1 with `c1xx: fatal error C1060: compiler is out of heap space` error.

I'd recommend try another approach: just change the `detect_fields_count_greedy` to fill an `bool init_succeeded[Last]`. This could be done in one go, by getting an index sequence from Last, and applying it to the function that returns true/false, depending on the construction success. After that a simple loop in constexpr could finish the job.

Something like that:
```
size_t detect_fields_count_greedy(index_sequence<Indexes...>) {
bool init_succeeded[Last] = { is_aggr_initable<T, Indexes≥(), ... };
for ( i = Last - 1; i > 0; --i) if init_succeeded [i] return i;
}
```

**Comment by runer112** [comment]
In the last CI run, 15 tasks failed with a compiler is out of heap space error. With the jobs running in parallel, it's hard to determine which tasks failed due to their own excessive memory usage and which were well-behaved, but a victim of running when another task consumed all the available memory. I pushed a temporary commit that I believe should disable testing in parallel. Could you please approve the CI build?

Regarding your suggestion, maybe I'm not understanding it correctly, but I don't see how it would help with performance. The crux of the performance issue is that the check of whether a type is constructible with N arguments, `enable_if_constructible_helper_t` (SFINAE), costs O(N) memory and time. In your suggestion, I believe this would make initializing `init_succeeded` cost O(Last^2) time and O(Last) memory.

My proposed changes aim to minimize the overall cost by minimizing the sum of all N checked. They effectively replace `sizeof(T)` with just the result (field count) in asymptotic performance analysis, which may be substantially lesser. Comparing to the current code (don't trust the current comments):

Case                             | Memory Before | Memory After | Time Before                   | Time After
-------------------------------- | ------------- | ------------ | ----------------------------- | -----------------------
`T` is an array                  | O(1)          | O(1)         | O(1)                          | O(1)
`T` is default-constructible     | O(sizeof(T))  | O(result)    | O(sizeof(T) * log(sizeof(T))) | O(result * log(result))
`T` is not default-constructible | O(sizeof(T))  | O(result)    | O(sizeof(T)^2)                | O(result^2)

**Comment by runer112** [comment]
The excessive compile time/memory usage issue has been fixed.

**Comment by runer112** [comment]
@apolukhin Is there anything more you'd like me to address?

The AppVeyor build succeeds now, and does so 5-10% faster than before despite there being three additional tests.

**Comment by apolukhin** [comment]
Sorry, I've misread you for the first time.

So here's how I see your changes (please fix me if I'm wrong):

T is default-constructible: you do exponentioal search for upper bound of fields count. It takes log(fields_count) + 1. After that you do a binary search that takes log(log(fields_count) + 1). The final complexity is log(fields_count) + 1 + log(log(fields_count) + 1).

The current implementation starts the binary search from sizeof(type) * CHAR_BIT. However, that CHAR_BIT multiplication is not necessary - we do not need to know the eact count of bitfields. Instead of that the binary search could work with sizeof(type)+1, and if we get the maximum value, then the type has bitfields.

Here comes the math. Your algorithm is better when
log(sizeof(type)+1) > log(fields_count) + 1 + log(log(fields_count) + 1)

which is 
log(sizeof(type)+1) > log(fields_count) + log(2) + log(log(fields_count) + 1)

which is 
sizeof(type)+1 > fields_count*2 * (log(fields_count) + 1)

sizeof(type) is equal to fields_count*avg_field_size. It gives us

fields_count*avg_field_size+1 > fields_count*2 * (log(fields_count) + 1)

Which is
avg_field_size+1/fields_count > 2 * log(fields_count) + 2

For fields count 16 the avg_field_size should be about 10 to make your algorithm better. For fields count 256 the avg_field_size should be about 18 to make your algorithm better.

For aggregates of ints, chronos, pointers or size_ts existing algorithm performs better. For aggregates of strings and vectors your algorithm performs better than the existing. 

I'd rather call it a tie. But the CHAR_BITS multiplocation should be removed.

T is not defsult constructible: your approach is defenetly superior. I'm worried about the cases, when the whole type is not aggregate initializable, because in that case I think your algo would run as long as the RAM os not exhausted and no diagnostic will be provided. Probably it is the reason, why github CI fails.

I'm also worried about compiler idiosyncrasies. Not all the compikers are listed in CI, so I'd rather stick to the existing, well tested algorithm, if it does not make a noticeable difference.

So, here's the plan:
* remove the CHAR_BITS multiplication
* for the second case: do the linear search for first non default constructible T, and then use the existing binary search with sizeof(T)+1 upper limit

**Comment by runer112** [comment]
Your understanding of the approach used in these changes is correct: exponential search followed by binary search. However, your performance analysis only counts "steps". Critically, it does not factor in the cost of checking at each step whether the type is constructible with N arguments, which is O(N) memory and time.

Factoring this in to the default-construble case, the exponential search worst case costs O(1 + 2 + 4 + ... + fields_count + 2 * fields_count) = O(4 * fields_count) to establish bounds separated only by a factor of 2. The best case costs O(1 + 2 + 4 + ... + fields_count + 1) = O(2 * fields_count).

Without exponential search and ignoring the possibility of bitfields, we start with a binary search over [0, sizeof(T)]. To establish bounds separated by a factor of 2, in the best case of an average field size in [1, 2] bytes, this only requires one check costing O(sizeof(T) / 2). This ranges from O(field_count / 2) to O(field_count), which is 1/4 to 1/2 of exponential search's best case cost.

However, the favorable comparison fades rather quickly. Once the average field size passes 4 bytes, establishing bounds separated by a factor of 2 requires (at least) three checks costing O(sizeof(T) / 2 + sizeof(T) / 4 + sizeof(T) / 8) = O(7/8 * sizeof(T)). The field count must be less than sizeof(T) / 4, so in terms of field count, the cost is at least O(7/8 * 4 * field_count) = O(7/2 * field_count). This is already nearly the worst case cost of exponential search, and it only gets worse for larger average field sizes.

In the worst case of a single field, the cost is O(sizeof(T) / 2 + sizeof(T) / 4 + sizeof(T) / 8 + ... 1) = O(sizeof(T)). This is sizeof(T) / 4 times exponential search's worst case cost. The cost factor is nearly unbounded as it simply grows with sizeof(T).

Considering that the best case cost factor of using binary search only is 1/4 and the worst case cost factor is nearly unbounded, I believe that that alone should be enough reason to use exponential search. Additionally, it seems that exponential search may be faster in "average" use, as evidenced by the AppVeyor builds with these changes being roughly 5% faster than builds without. And as a reminder, really poor performance cases with binary search aren't just a theoretical possibility that won't happen in practice. I went through the effort to make and propose these changes specifically because I ran into such a poorly performing case where these changes made a huge difference (https://github.com/boostorg/pfr/pull/120#issuecomment-1396202005).

Regarding the latest CI failure, I haven't had much time to look into it yet. I was hoping that it wouldn't be necessary to do this because it feels unclean, but perhaps it's wise to add a sanity check to the exponential search to halt it if it somehow exceeds `sizeof(T) * CHAR_BIT`.

**Comment by apolukhin** [comment]
> Critically, it does not factor in the cost of checking at each step whether the type is constructible with N arguments, which is O(N) memory and time.

It should not take O(N). `std::make_index_sequence` was a bottleneck in so many cases, that the compiler developers made it an intrinsic (for example https://github.com/microsoft/STL/blob/73924c1920af92899f7582cd904ea819b9db35bc/stl/inc/type_traits#L34). So getting the indexes is O(1), variadic pack expansion is also close to O(1), `ubiq_*ref_constructor` is not a template and its constructors do not consume time/memory. As a result the check should take about O(1).

Please, check that your compiler supports the builtin and that it is properly detected in https://github.com/boostorg/pfr/blob/28bd7f541f7a7632f4fe52e30d06a121e7cb1f65/include/boost/pfr/detail/make_integer_sequence.hpp

**Comment by runer112** [comment]
I noticed that someone else (@rressi-at-globus) appears to have discovered these changes and found the changes to be useful enough to make their own attempt to get these changes upstreamed with #179. I resynchronized my branch with `develop` to hopefully allow this PR to proceed, now with third-party corroboration that these changes are beneficial.

**Comment by runer112** [comment]
I re-ran the tests that I previously ran in https://github.com/boostorg/pfr/pull/120#issuecomment-1420993441 to once again demonstrate that `std::make_index_sequence` is not constant cost with respect to the size of the sequence, but now with more recent versions of clang, gcc, and msvc just in case something got optimized. But my findings seem to be the same as before: `std::make_index_sequence` is not constant cost with respect to the size of the sequence.

1. https://godbolt.org/z/PKqes5h1M: `std::make_index_sequence<10'000'000>` gave me the result across multiple retries of resource exhaustion for all 3 compilers.
2. https://godbolt.org/z/9jEGEG776: `std::make_index_sequence<1'000'000>` gave me best-case results of clang completing in 0.9 seconds, gcc completing in 2.6 seconds, and msvc never completing due to resource exhaustion.
3. https://godbolt.org/z/PKqes5h1M: `std::make_index_sequence<100'000>` gave me best-case results of clang completing in 0.3 seconds, gcc completing in 0.5 seconds, and msvc completing in 1.3 seconds.

**Comment by runer112** [comment]
@apolukhin: Here are direct reponses to snippets from your comment, https://github.com/boostorg/pfr/pull/120#issuecomment-1415440892. I apologize for not doing this earlier.

---

> I'm worried about the cases, when the whole type is not aggregate initializable, because in that case I think your algo would run as long as the RAM os not exhausted and no diagnostic will be provided.

You were right, there was a problem. I added tests for a default-constructible type accepting 0 or more arguments and a non-default-constructible type accepting 1 or more arguments in e80f7aa951676d5c25d3d264b87d63eebe505d5f. These tests should now pass with the changes from dd1ae1c907a6181ff2eff32cf1091450b9950df1.

---

> I'd rather stick to the existing, well tested algorithm, if it does not make a noticeable difference

It does make a noticeable difference for large types.

Practical evidence:

1. https://github.com/boostorg/pfr/pull/120#issuecomment-1396202005
2. https://github.com/boostorg/pfr/pull/120#issuecomment-2332159961

Theoretical evidence:

1. https://github.com/boostorg/pfr/pull/120#issuecomment-1416529425
2. https://github.com/boostorg/pfr/pull/120#issuecomment-2332496989

---

> **Here's the plan:**
> 
> * remove the CHAR_BITS multiplication
> * for the second case: do the linear search for first non default constructible T, and then use the existing binary search with sizeof(T)+1 upper limit

As you're aware, the current implementation performs binary search with an upper bound of `sizeof(T) * CHAR_BIT`. For large types, the first few iterations could carry a significant cost, perhaps even resulting in resource exhaustion. This could maybe warrant removing the multiplication by `CHAR_BIT` at the cost of losing support for types with dense bitfields (average field size < 1 byte).

The proposed implementation would perform binary search with an upper bound that is no more than twice the field count. This would not carry any unnecessary significant cost. In this regard, removing the multiplication by `CHAR_BIT` should not be warranted.

There is however a preexisting case of concern: large types that are not aggregate-initializable in environments that do not provide `std::is_aggregate`. With both the existing and the proposed implementation, this may result in searching all the way up to `sizeof(T) * CHAR_BIT` just to eventually encounter resource exhaustion or produce an error stating that the type is not aggregate-initializable. But `std::is_aggregate` is required by the C++17 standard, environments without it will only become rarer with time, and using a non-aggregate type with pfr is a misuse to begin with. So with the proposed implementation, removing the multiplication by `CHAR_BIT` solely to prevent poor behavior in the case of misuse with a dated C++ environment, while also at the cost of losing support for types with dense bitfields, doesn't seem worth it to me. But if I've forgotten to account for another problematic case or if you otherwise disagree and still want to the multiplication by `CHAR_BIT` to be removed, let me know.

**Comment by apolukhin** [comment]
Still fails the tests

**Comment by runer112** [comment]
I'm looking at the output of the failed runs now, but there seems to be a lot of unstructured text to comb through... Any tips as to what to look at?

**Comment by runer112** [comment]
After my best attempts to search for failures in the log files, I could not spot any test failures, but I did spot a couple of Boost Inspection Report problems. These should now be addressed.

**Comment by runer112** [comment]
Bleh, one more failing CI job. It's late for me today, but I can investigate that tomorrow.

Also, it appears that the coverage report didn't work because I merged `develop` back into this branch rather than rebasing. Would you like me to rebase instead?

**Comment by runer112** [comment]
I have implemented a workaround for an MSVC issue that caused failures in [windows (msvc-14.3, 20,latest, 64, windows-2022, -j1)](https://github.com/boostorg/pfr/actions/runs/10855330620/job/30358979387?pr=120#logs). It's hard for me to tell from the output if this was the only issue, but I'm hopeful that it was.

**Comment by runer112** [comment]
I've made additional improvements that should remove any unnecessary trial initializations of array types. This means that the cost of calculating the field count of an array is now be constant. This is tested here:

https://github.com/runer112/pfr/blob/cbc57cc28758c36c77b3fbe4e4ef7fb8e98f5ae8/test/core/run/huge_count.cpp#L53

I've also finally added testing of the core improvement that this PR provides, which is that the cost of calculating the field count of an object is not proportional to the object's size. This is tested here:

https://github.com/runer112/pfr/blob/cbc57cc28758c36c77b3fbe4e4ef7fb8e98f5ae8/test/core/run/huge_count.cpp#L55-L56

**Comment by runer112** [comment]
The latest CI run revealed 3 embarrassingly simple issues that slipped in last week while making robustness and testing improvements. I have pushed commits to address these issues.

File: include/boost/pfr/detail/fields_count.hpp [hunk 1]
```diff
@@ -74,25 +83,28 @@ template <class T> struct is_single_field_and_aggregate_initializable<1, T>: std
...
+template <class T, std::size_t N, class /*Enable*/ = void>
+struct is_aggregate {
+    static constexpr bool value =
+           std::is_empty<T>::value
+        || std::is_array<T>::value
+    ;
+};
+
 template <class T, std::size_t N>
-struct is_aggregate_initializable_n {
```
Comment (apolukhin) [inline include/boost/pfr/detail/fields_count.hpp:82]: There were changes in the C++ standard about aggregates and aggregate initialization. I then the old name is more precise

File: include/boost/pfr/detail/fields_count.hpp [hunk 2]
```diff
@@ -181,143 +227,213 @@ constexpr std::size_t detect_fields_count(detail::one_element_range, long) noexc
...
-    return N;
+    return Begin + I - 1;
 }
 
-template <class T, std::size_t N>
-constexpr std::size_t detect_fields_count_greedy_remember(int) noexcept {
-    return 0;
+template <class T, std::size_t Begin, std::size_t I>
+constexpr auto fields_count_upper_bound(long, long) noexcept
+    -> std::enable_if_t<(Begin + I > fields_count_upper_bound_loose<T>()), std::size_t>
```
Comment (apolukhin) [inline include/boost/pfr/detail/fields_count.hpp:254]: Add namespace to function call to avoid ADL

**Review by apolukhin — CHANGES_REQUESTED** [review]

File: include/boost/pfr/detail/fields_count.hpp [hunk 3]
```diff
@@ -181,143 +227,213 @@ constexpr std::size_t detect_fields_count(detail::one_element_range, long) noexc
...
-//#if !BOOST_PFR_USE_CPP17
-//    static_assert(
-//        std::is_standard_layout<type>::value,   // Does not return `true` for structs that have non standard layout members.
-//        "Type must be aggregate initializable."
-//    );
-//#endif
-
-#if defined(_MSC_VER) && (_MSC_VER <= 1920)
-    // Workaround for msvc compilers. Versions <= 1920 have a limit of max 1024 elements in template parameter pack
-    constexpr std::size_t max_fields_count = (sizeof(type) * CHAR_BIT >= 1024 ? 1024 : sizeof(type) * CHAR_BIT);
```
Comment (apolukhin) [inline include/boost/pfr/detail/fields_count.hpp:304]: Move this workaround to fields_count_upper_bound_loose

File: include/boost/pfr/detail/fields_count.hpp [hunk 4]
```diff
@@ -181,143 +227,213 @@ constexpr std::size_t detect_fields_count(detail::one_element_range, long) noexc
...
+constexpr auto fields_count_upper_bound(long, long) noexcept
+    -> std::enable_if_t<(Begin + I > fields_count_upper_bound_loose<T>()), std::size_t>
+{
+    static_assert(
+        !detail::is_initializable<T, fields_count_upper_bound_loose<T>() + 1>(1L),
+        "====================> Boost.PFR: Types with user specified constructors (non-aggregate initializable types) are not supported.");
+    return fields_count_upper_bound_loose<T>();
+}
+
+template <class T, std::size_t Begin, std::size_t I>
```
Comment (apolukhin) [inline include/boost/pfr/detail/fields_count.hpp:262]: Looks like Begin is always 0 in all the fields_count_upper_bound functions. Just remove that template parameter

**Review by runer112 — COMMENTED** [review]
Thank you for your code review, @apolukhin. I have read all of your comments. I replied to 3 comments that I wasn't sure how to proceed with and will await your response. All of the other comments make sense and I will address them ASAP.

File: include/boost/pfr/detail/fields_count.hpp (hunk 1 above)
Comment (runer112) [reply include/boost/pfr/detail/fields_count.hpp:82]: [Restore "is_aggregate_initializable_n" name](https://github.com/boostorg/pfr/pull/120/commits/d222d982f9e8481e48d45cc39ec6ce58cff14cb1)

File: include/boost/pfr/detail/fields_count.hpp [hunk 5]
```diff
@@ -15,6 +15,7 @@
 #ifdef BOOST_PFR_HAS_STD_MODULE
 import std;
 #else
+#include <algorithm>    // min
```
[the remark this answers is not in the document]
Comment (runer112) [reply include/boost/pfr/detail/fields_count.hpp]: [Avoid including \<algorithm\>](https://github.com/boostorg/pfr/pull/120/commits/ae8ce8dfe32842250e84123a5c7db2e5a04e8fac)

File: include/boost/pfr/detail/fields_count.hpp (hunk 2 above)
Comment (runer112) [reply include/boost/pfr/detail/fields_count.hpp]: [Prevent ADL throughout fields_count.hpp](https://github.com/boostorg/pfr/pull/120/commits/ff14705fac6bf2c1f697fe776f87e171b147d622)

File: include/boost/pfr/detail/fields_count.hpp [hunk 6]
```diff
@@ -181,143 +227,213 @@ constexpr std::size_t detect_fields_count(detail::one_element_range, long) noexc
...
-constexpr std::size_t detect_fields_count_greedy_remember(int) noexcept {
-    return 0;
+template <class T, std::size_t Begin, std::size_t I>
+constexpr auto fields_count_upper_bound(long, long) noexcept
+    -> std::enable_if_t<(Begin + I > fields_count_upper_bound_loose<T>()), std::size_t>
+{
+    static_assert(
+        !detail::is_initializable<T, fields_count_upper_bound_loose<T>() + 1>(1L),
+        "====================> Boost.PFR: Types with user specified constructors (non-aggregate initializable types) are not supported.");
+    return fields_count_upper_bound_loose<T>();
```
[the remark this answers is not in the document]
Comment (runer112) [reply include/boost/pfr/detail/fields_count.hpp]: [Prevent ADL throughout fields_count.hpp](https://github.com/boostorg/pfr/pull/120/commits/ff14705fac6bf2c1f697fe776f87e171b147d622)

File: include/boost/pfr/detail/fields_count.hpp [hunk 7]
```diff
@@ -181,143 +227,213 @@ constexpr std::size_t detect_fields_count(detail::one_element_range, long) noexc
...
-    constexpr std::size_t fields_count_small_range = detail::detect_fields_count_greedy<T, small_range_begin, small_range_last>(
-        detail::is_one_element_range<small_range_begin, small_range_last>{}
+template <class T, std::size_t Begin = 1, std::size_t Result>
+constexpr std::size_t fields_count_lower_bound_unbounded(int, size_t_<Result>) noexcept {
+    return Result;
+}
+
+template <class T, std::size_t Begin = 1>
+constexpr auto fields_count_lower_bound_unbounded(long, size_t_<0> = {}) noexcept
+    -> std::enable_if_t<(Begin >= fields_count_upper_bound_loose<T>()), std::size_t>
```
[the remark this answers is not in the document]
Comment (runer112) [reply include/boost/pfr/detail/fields_count.hpp]: [Prevent ADL throughout fields_count.hpp](https://github.com/boostorg/pfr/pull/120/commits/ff14705fac6bf2c1f697fe776f87e171b147d622)

File: include/boost/pfr/detail/fields_count.hpp [hunk 8]
```diff
@@ -181,143 +227,213 @@ constexpr std::size_t detect_fields_count(detail::one_element_range, long) noexc
...
+{
+    static_assert(
+        detail::is_initializable<T, fields_count_upper_bound_loose<T>()>(1L),
```
[the remark this answers is not in the document]
Comment (runer112) [reply include/boost/pfr/detail/fields_count.hpp]: [Prevent ADL throughout fields_count.hpp](https://github.com/boostorg/pfr/pull/120/commits/ff14705fac6bf2c1f697fe776f87e171b147d622)

File: include/boost/pfr/detail/fields_count.hpp [hunk 9]
```diff
@@ -181,143 +227,213 @@ constexpr std::size_t detect_fields_count(detail::one_element_range, long) noexc
...
+        "====================> Boost.PFR: Type must be aggregate initializable.");
+    return fields_count_upper_bound_loose<T>();
```
[the remark this answers is not in the document]
Comment (runer112) [reply include/boost/pfr/detail/fields_count.hpp]: [Prevent ADL throughout fields_count.hpp](https://github.com/boostorg/pfr/pull/120/commits/ff14705fac6bf2c1f697fe776f87e171b147d622)

File: include/boost/pfr/detail/fields_count.hpp [hunk 10]
```diff
@@ -181,143 +227,213 @@ constexpr std::size_t detect_fields_count(detail::one_element_range, long) noexc
...
+}
+
+template <class T, std::size_t Begin = 1>
+constexpr std::size_t fields_count_lower_bound_unbounded(int, size_t_<0> = {}) noexcept
+{
+    constexpr std::size_t last = (std::min)(Begin * 2, fields_count_upper_bound_loose<T>()) - 1;
```
[the remark this answers is not in the document]
Comment (runer112) [reply include/boost/pfr/detail/fields_count.hpp]: [Prevent ADL throughout fields_count.hpp](https://github.com/boostorg/pfr/pull/120/commits/ff14705fac6bf2c1f697fe776f87e171b147d622)

File: include/boost/pfr/detail/fields_count.hpp [hunk 11]
```diff
@@ -181,143 +227,213 @@ constexpr std::size_t detect_fields_count(detail::one_element_range, long) noexc
...
+    constexpr std::size_t begin = detail::fields_count_lower_bound_unbounded<T>(1L);
+    return detail::fields_count_binary_search_unbounded<T, begin>();
 }
 
 ///////////////////// Returns fields count
 template <class T>
 constexpr std::size_t fields_count() noexcept {
     using type = std::remove_cv_t<T>;
 
+    constexpr bool precondition1 = detail::is_complete<type>::value;
```
[the remark this answers is not in the document]
Comment (runer112) [reply include/boost/pfr/detail/fields_count.hpp]: - [Revert "Prevent unnecessary triggering of last-resort assertion"](https://github.com/boostorg/pfr/pull/120/commits/13f29c635f89d17a65abde5e59a05e7d4dfb9861)
- [Revert "Add complete type precondition and condition evaluation on preconditions"](https://github.com/boostorg/pfr/pull/120/commits/ab3621c45d4cee56fd70c61a9ae162fcd7608906)
- [Add back complete type precondition](https://github.com/boostorg/pfr/pull/120/commits/d6677deeaf802726a36e875fdb5707d618a0d533)

File: include/boost/pfr/detail/fields_count.hpp (hunk 3 above)
Comment (runer112) [reply include/boost/pfr/detail/fields_count.hpp:304]: [Stop at expected compiler limitation values during exponential search](https://github.com/boostorg/pfr/pull/120/commits/68e6abaa255c0d594da725a961ccad4c467dad85)

File: include/boost/pfr/detail/fields_count.hpp (hunk 4 above)
Comment (apolukhin) [reply include/boost/pfr/detail/fields_count.hpp]: Ok, then it is fine

File: include/boost/pfr/detail/fields_count.hpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (apolukhin) [reply include/boost/pfr/detail/fields_count.hpp]: Oh, then the change is fine, but let's make it in a separate PR

File: include/boost/pfr/detail/fields_count.hpp [hunk 12]
```diff
@@ -30,6 +31,12 @@ import std;
 
 namespace boost { namespace pfr { namespace detail {
 
+///////////////////// min without including <algorithm>
+template <class T>
+constexpr const T &min(const T &a, const T &b) {
```
Comment (apolukhin) [inline include/boost/pfr/detail/fields_count.hpp:36]: Just remove all the references and pass std::size_t by value

File: include/boost/pfr/detail/fields_count.hpp [hunk 13]
```diff
@@ -170,9 +211,26 @@ using is_one_element_range = std::integral_constant<bool, Begin == Last>;
 using multi_element_range = std::false_type;
 using one_element_range = std::true_type;
 
-///////////////////// Non greedy fields count search. Templates instantiation depth is log(sizeof(T)), templates instantiation count is log(sizeof(T)).
+
+///////////////////// Fields count next expected compiler limitation
+constexpr std::size_t fields_count_compiler_limitation_next(std::size_t n) noexcept {
+#if defined(_MSC_VER) && (_MSC_VER <= 1920)
+    if (n < 1024)
```
Comment (apolukhin) [inline include/boost/pfr/detail/fields_count.hpp:218]: Wrong comparison?

**Review by apolukhin — CHANGES_REQUESTED** [review]
Looks almost perfectly!

Added a few comments

**Review by runer112 — COMMENTED** [review]
I have replied to your new comments and will await your responses. I have also pushed a simple fix for what seems to be the only remaining issue identified by CI: [Address violation of Boost min/max guidelines](https://github.com/boostorg/pfr/pull/120/commits/fe6f150a32c3ed6e6e56ae42b650ad3ac4e7952b).

**Comment by apolukhin** [comment]
Awesome work!

Many thanks for it!

**Comment by runer112** [comment]
> Awesome work!
> 
> Many thanks for it!

And thank you for hanging in there with me through the long process to ensure that these changes meet the high standards expected for a widely-used library.

Although I hope this should never happen, if any concerns are raised about this set of changes in the future, feel free to contact me for support.

**Pull request merged**